flask_potion/contrib/alchemy/manager.py

```diff
@@ -256,88 +256,87 @@ def _query_get_first(self, query):
         except NoResultFound:
             raise IndexError()
 
-    def create(self, properties, commit=True):
+    def create(self, properties):
         # noinspection properties
         item = self.model(**properties)
 
         before_create.send(self.resource, item=item)
-
         session = self._get_session()
-
-        try:
-            session.add(item)
-            self.commit_or_flush(commit)
-        except IntegrityError as e:
-            session.rollback()
-
-            if hasattr(e.orig, 'pgcode'):
-                if e.orig.pgcode == "23505":  # duplicate key
-                    raise DuplicateKey(detail=e.orig.diag.message_detail)
-
-            if current_app.debug:
-                raise BackendConflict(
-                    debug_info=dict(
-                        exception_message=str(e), statement=e.statement, params=e.params
+        with session.begin_nested():
```

Comment:
To summarize my understanding.

We want to avoid having `Potion` (that uses `flask-sqlalchemy`) committing the current transaction when models are created/updated/removed. The previous code is committing because of the default parameter `commit`, e.g `create(self, properties, commit=True)`.

At this point we could simply set `commit=False` and we would [flush](https://docs.sqlalchemy.org/en/13/orm/session_api.html#sqlalchemy.orm.session.Session.flush) the session which keeps the current transactional concept, except if the operation fails (in this case the transaction is rollbacked).

In addition, [#2028](https://github.com/infarm/kompost/pull/2028) rollbacks the transaction in `Potion`'s exception handler.

If I understand correctly, the only reason to add a savepoint (`session.begin_nested()`) is for our tests as we do not want to rollback the data preparation step.

We can have different failure scenarios (where the transaction has to be handled correctly):
1. the model cannot be created because a validation steps fails (e.g try to create a HarvestRecord)
1. database operations fail during a `flush`

I want to write the different flow diagrams in the production and test cases.

- Production: Model Validation Error
   1. Open a transaction `S0`
   1. We try to create a model, e.g `HarvestRecord`, but a `ModelValidationError` is raised
   1. Potion `_exception_handler()` catches this exception and issues a `Rollback`, back to `S0`
- Production: Persistence error
   1. Open a transaction (`S0`)
   1. We create a model, a `savepoint` is created within the transaction (`S1`)
   1. `session.flush()` fails, the transaction is rolled back, we are back to `S1` and an exception is raised
   1. Potion `_exception_handler()` catches the `flush` exception and issues a `Rollback`, back to `S0`

For the tests, I am unsure what exactly happens. Right now I do not see why `begin_nested` is needed.